chain_of_trust.go

```
@@ -0,0 +1,172 @@
+	for _, artifact := range task.Artifacts {
+		switch a := artifact.(type) {
+		case S3Artifact:
+			// make sure SHA256 is calculated
+			hash, err := calculateHash(a)
+			if err != nil {
+				return err
+			}
+			artifactHashes = append(
+				artifactHashes,
+				ArtifactHash{
+					Name: a.CanonicalPath,
+					Hash: "sha256:" + hash,
+				})
+		}
+	}
+
+	cotCert := &ChainOfTrustCertificate{
+		Artifacts:   artifactHashes,
+		Task:        task.Definition,
+		TaskID:      task.TaskID,
+		RunID:       task.RunID,
+		WorkerGroup: config.WorkerGroup,
+		WorkerID:    config.WorkerID,
+		Extra: CotExtra{
+			PublicIPAddress:  config.PublicIP.String(),
+			PrivateIPAddress: config.PrivateIP.String(),
+			InstanceID:       config.InstanceID,
+			InstanceType:     config.InstanceType,
+			Region:           config.Region,
+		},
+	}
+
+	certBytes, err := json.MarshalIndent(cotCert, "", "  ")
+	if err != nil {
+		return err
+	}
+	// separate signature from json with a new line
+	certBytes = append(certBytes, '\n')
+
+	in := bytes.NewBuffer(certBytes)
+	out, err := os.Create(signedCert)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Write(certBytes)
+
+	privKeyFile, err := os.Open(config.SigningKeyLocation)
+	if err != nil {
+		return err
+	}
+	defer privKeyFile.Close()
+	entityList, err := openpgp.ReadArmoredKeyRing(privKeyFile)
+	if err != nil {
+		return err
+	}
+	privKey := entityList[0]
+	openpgp.ArmoredDetachSign(out, privKey, in, nil)
+	out.Close()
+	err = task.uploadLog("public/logs/certified.log.gpg")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cot *ChainOfTrust) Created(task *TaskRun) error {
+	return nil
+}
+
+func appendFileContents(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return
+	}
+	defer in.Close()
+	out, err := os.OpenFile(dst, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
```

Isn't it `!=`? Anyway, won't it hide an error from `out.Sync`?